src/kinesis.js

```diff
@@ -1,6 +1,224 @@
+
+const KINESIS_IMAGE = 'localstack/localstack:latest';
+
+let kinesisStreams = [];
+
+function streams (streams) {
+  kinesisStreams = streams.map(StreamName => ({
+    StreamName,
+    ShardCount: 1
+  }));
+}
+
+function getStreamName (streamName, uniqueIdentifier) {
+  return uniqueIdentifier ? `${streamName}-${uniqueIdentifier}` : streamName;
+}
+
+async function createStreams (kinesisClient, uniqueIdentifier) {
+  const failedProvisons = [];
+  await Promise.all(
+    kinesisStreams.map(async stream => {
+      const newStream = cloneDeep(stream);
+      const StreamName = getStreamName(newStream.StreamName, uniqueIdentifier);
+      newStream.StreamName = StreamName;
+
+      try {
+        await kinesisClient.createStream(newStream).promise();
+        await kinesisClient.waitFor('streamExists', { StreamName }).promise();
+      } catch (err) {
+        failedProvisons.push(StreamName);
+        console.error(`Failed to create stream "${StreamName}"`, err);
+      }
+    })
+  );
+
+  if (failedProvisons.length) {
+    throw new Error(`Failed to create streams: ${failedProvisons.join(', ')}`);
+  }
+}
+
+async function destroyStreams (kinesisClient, uniqueIdentifier) {
+  const failedDeletions = [];
+  const { StreamNames } = await kinesisClient.listStreams().promise();
+  const schemaTableNames = kinesisStreams
+    .map(({ StreamName }) => getStreamName(StreamName, uniqueIdentifier));
+  const tablesToDestroy = StreamNames
+    .filter(name => schemaTableNames.includes(getStreamName(name, uniqueIdentifier)));
+
+  await Promise.all(
+    tablesToDestroy
+      .map(async StreamName => {
+        try {
+          await kinesisClient.deleteStream({ StreamName }).promise();
+          await kinesisClient.waitFor('streamNotExists', { StreamName }).promise();
+        } catch (err) {
+          failedDeletions.push(StreamName);
+          console.error(`Failed to destroy stream "${StreamName}"`, err);
+        }
+      })
+  );
+
+  if (failedDeletions.length) {
+    throw new Error(`Failed to destroy streams: ${failedDeletions.join(', ')}`);
+  }
+}
+
+function buildStreamNameMapping (uniqueIdentifier) {
+  return fromPairs(kinesisStreams.map(({StreamName}) => {
+    return [StreamName, getStreamName(StreamName, uniqueIdentifier)];
+  }));
+}
+
+function buildConfigFromConnection (connection) {
+  return {
+    credentials: new AWS.Credentials(connection.accessKey, connection.secretAccessKey),
+    endpoint: new AWS.Endpoint(connection.url),
+    region: connection.region,
+    httpOptions: {
+      timeout: 3000
+    },
+    maxRetries: 3
+  };
+}
+
+async function waitForKinesisToBeReady (kinesisClient) {
+  await promiseRetry(async function (retry, retryNumber) {
+    try {
+      await kinesisClient.listStreams({Limit: 1}).promise();
+    } catch (error) {
+      retry(new NestedError(`Kinesis is still not ready after ${retryNumber} connection attempts`, error));
+    }
+  }, {factor: 1, minTimeout: 100, retries: 100});
```

Comment:
This seems pretty aggressive. Was there a particular reason for 100ms timeouts with 100 attempts? Docker operations tend to be pretty heavy weight so I would expect very few things to succeed in less than 100ms. It might be nicer to the host to do something more like 10 attempts at 1s. Also since the AWS config uses a 3s timeout with 3 retries this seems like it could potentially create a snowball of `listStreams` requests.

Reply:
I agree, it is pretty fast.  I think changing it to run every 500ms, then giving it 10 seconds to finish still should be good.  The minTimeout means how long to wait between attempts, as a minimum.  There is no maxTimeout so it will wait for the promise to fail before trying again.  I will update DynamoDB to use the same config.